### PR TITLE
Feat/restrict metadata

### PR DESCRIPTION
**Pull request by alvinsw**
Added access restriction functionality based on the:
1. For endpoints `/crate`, `/entity`, `/file`: responds with 403 when `access.metadata` returned by accessTransformer function is false.
2. For endpoints `/search`, `/entities`, `/files`: Use the resolveValidLicenses helper function (if it is specified in the options) to retrieve a list of valid licenses granted to the logged-in user and use the list to filter the search query.

### CODE REVIEW

**Comment by alvinsw** [comment]
@johnf coverage has been fix, can you review please.

File: src/test/helpers/fastify.ts [hunk 1]
```diff
@@ -25,3 +26,23 @@ export const fastifyBefore = async () => {
 export const fastifyAfter = async () => {
   await fastify.close();
 };
+
+export const RestrictedAccessTransformer: AccessTransformer = (entity) => ({
+  ...entity,
+  access: {
+    metadata: false,
+    content: false,
+    metadataAuthorizationUrl: '',
```
Comment (johnf) [inline src/test/helpers/fastify.ts]: Can you add some dummy URLs here? I'm worried about JavaScript's truthiness logic and this hitting some strange edge case,

File: src/routes/crate.test.ts [hunk 2]
```diff
@@ -421,3 +426,69 @@ describe('Crate Route', () => {
...
+
+      const response = await fastify.inject({
+        method: 'HEAD',
+        url: `/entity/${encodeURIComponent('http://example.com/entity/file.wav')}/rocrate`,
+      });
+      const body = JSON.parse(response.body) as { error: { code: string; message: string } };
+
+      expect(response.statusCode).toBe(403);
+      expect(body.error.code).toBe('FORBIDDEN');
+      expect(mockRoCrateHandler.get).not.toHaveBeenCalled();
```
Comment (johnf) [inline src/routes/crate.test.ts]:
```suggestion
      expect(mockRoCrateHandler.head).not.toHaveBeenCalled();
```

File: src/routes/entity.ts [hunk 3]
```diff
@@ -49,6 +49,8 @@ const entity: FastifyPluginAsync<EntityRouteOptions> = async (fastify, opts) =>
         };
         const authorisedEntity = await accessTransformer(standardEntity, { request, fastify });
 
+        if (!authorisedEntity.access.metadata)
```
Comment (johnf) [inline src/routes/entity.ts]: Can you add `{}` around the block.

File: src/routes/entities.test.ts [hunk 4]
```diff
@@ -318,3 +318,62 @@ describe('Entities Route', () => {
     });
   });
 });
+
+describe('Entities Route with License Filtering', () => {
+  let hasLicense = true;
+  async function resolveValidLicenses() {
+    if (hasLicense) return ['https://creativecommons.org/licenses/by/4.0/'];
```
Comment (johnf) [inline src/routes/entities.test.ts]: {} and multiline block here please

File: src/app.ts [hunk 5]
```diff
@@ -83,6 +84,7 @@ export type Options = {
   fileTransformers?: FileTransformer[];
   fileHandler: FileHandler;
   roCrateHandler: RoCrateHandler;
+  resolveValidLicenses?: (opt: TransformerContext) => Promise<string[]>;
```
Comment (johnf) [inline src/app.ts]: Let's import the entire type rather than generate it here. It will be easier for consumers of the API to define their own functions.
```suggestion
  resolveValidLicenses?: LicenseResolver,
```

File: src/routes/entities.test.ts [hunk 6]
```diff
@@ -318,3 +318,62 @@ describe('Entities Route', () => {
     });
   });
 });
+
+describe('Entities Route with License Filtering', () => {
+  let hasLicense = true;
+  async function resolveValidLicenses() {
```
Comment (johnf) [inline src/routes/entities.test.ts]: Type this so we don't need the ts-expect-error below

File: src/routes/entities.ts [hunk 7]
```diff
@@ -54,6 +55,12 @@ const entities: FastifyPluginAsync<EntitiesRouteOptions> = async (fastify, opts)
           };
         }
 
+        if (resolveValidLicenses) {
+          where.metadataLicenseId = {
+            in: (await resolveValidLicenses({ request, fastify })) || [],
```
Comment (johnf) [inline src/routes/entities.ts]: the `|| []` shouldn't be necessary. We should require `resolveValidLicenses` to always return an array.

File: src/routes/crate.ts [hunk 8]
```diff
@@ -37,9 +39,17 @@ const crate: FastifyPluginAsync<CrateRouteOptions> = async (fastify, opts) => {
         if (!entity) {
           return reply.code(404).send(createNotFoundError('The requested entity was not found', id));
         }
+        const standardEntity = {
```
Comment (johnf) [inline src/routes/crate.ts]: Any reason not to use the `refMap` approach here as we do elsewhere?

File: src/routes/crate.ts [hunk 9]
```diff
@@ -75,6 +85,15 @@ const crate: FastifyPluginAsync<CrateRouteOptions> = async (fastify, opts) => {
           return reply.code(404).send(createNotFoundError('The requested entity was not found', id));
         }
 
+        const standardEntity = {
```
Comment (johnf) [inline src/routes/crate.ts]: As above

File: src/routes/files.ts [hunk 10]
```diff
@@ -48,18 +56,28 @@ const files: FastifyPluginAsync<FilesRouteOptions> = async (fastify, opts) => {
...
+        );
         // Apply transformers to each entity: base -> access -> additional
         const filesWithAccess = await Promise.all(
           dbFiles.map(async (dbFile) => {
-            const standardFile = baseFileTransformer(dbFile);
-            const authorisedFile = await fileAccessTransformer(standardFile, { request, fastify });
+            const entity = {
+              ...dbFile.entity,
+              ...baseFileTransformer(dbFile),
+              memberOf: dbFile.entity.memberOf ? (refMap.get(dbFile.entity.memberOf) ?? null) : null,
```
Comment (johnf) [inline src/routes/files.ts]: Why did this code change? Is this fixing a bug?

File: src/routes/search.ts [hunk 11]
```diff
@@ -60,8 +62,12 @@ const search: FastifyPluginAsync<SearchRouteOptions> = async (fastify, opts) =>
       },
     },
     async (request, reply) => {
-      const { searchType, query, filters, boundingBox, geohashPrecision, limit, offset, sort, order } = request.body;
-
+      const { searchType, query, boundingBox, geohashPrecision, limit, offset, sort, order } = request.body;
+      let filters = request.body.filters;
+      if (resolveValidLicenses) {
+        filters = filters || {};
+        filters.metadataLicenseId = (await resolveValidLicenses({ request, fastify })) || [];
```
Comment (johnf) [inline src/routes/search.ts]: I think this is right but we should have a quick discussion on Thursday.
Is it right that even for files we filter here by the metadataLicense (I think it is) but I think we need to document the rationale for this if we havne't already in the README.md

File: src/transformers/default.ts [hunk 12]
```diff
@@ -78,12 +78,14 @@ export type StandardFile = {
   size: number;
 };
 
+export type FileEntity = StandardFile & StandardEntity;
```
Comment (johnf) [inline src/transformers/default.ts]: This doesn't logically make sense. Why is it needed?

File: src/transformers/default.ts [hunk 13]
```diff
@@ -166,7 +168,7 @@ export const baseFileTransformer = (file: File): StandardFile => ({
  * });
  * ```
  */
-export const AllPublicFileAccessTransformer = (file: StandardFile): AuthorisedFile => ({
+export const AllPublicFileAccessTransformer = (file: FileEntity): AuthorisedFile => ({
```
Comment (johnf) [inline src/transformers/default.ts]: This feels like a breaking change in the API. Why is this necessary?

**Review by johnf — CHANGES_REQUESTED** [review]
- README.md needs updates to explain the new functionality

I'm not sure that this approach is general enough.
- resolveValidLicenses doesn't know if this is a meta or content license lookup
- What if I wanted to add other restrictions e.g. maybe I have special metadata in the entity I want to match against.

Let's discuss on Thursday I think we may need an approach where you can pass in a more complex Prisma element rather than just a list of licenses.

**Comment by johnf** [comment]
@alvinsw  I also asked Claude to review the PR. It picked up somethings I've mentioned but also some other items worth looking at

## Standards

### Hard violations (documented standards)

README.md — public API surface undocumented. resolveValidLicenses (src/app.ts:87) is a new consumer-facing Options field. README documents every other option (accessTransformer, queryBuilderClass, fileHandler, …); this one gets nothing. CLAUDE.md: "published npm library … be mindful of the public API surface".

README.md stale — "Applied Routes" (README:489-495). "The /file/:id endpoint … uses the fileHandler system, not file transformers." src/routes/file.ts now runs fileAccessTransformer on both GET and HEAD. Same section's "File Transformation Pipeline" (README:481-487) no longer describes stage 1: files.ts feeds the access transformer {...dbFile.entity,
  ...baseFileTransformer(dbFile)}, not the base transformer's output.

CLAUDE.md pipeline — src/routes/crate.ts:42-46, 88-93 skips the base transformer. It hand-builds {...entity, memberOf: {id: entity.memberOf || '', name: ''}} from the raw Prisma row, so consumer accessTransformers receive DB fields (meta, createdAt) and a fabricated name: '' instead of resolved references. entity.ts/search.ts use baseEntityTransformer + resolveEntityReferences.

Semver. FileAccessTransformer and AllPublicFileAccessTransformer change input from StandardFile to FileEntity, and AuthorisedFile gains all StandardEntity fields — consumer implementations that construct an AuthorisedFile break. Commits are feat:, no !/BREAKING CHANGE footer (cf. aa5c6b7 feat!:).

### Baseline smells (judgement calls)

- Duplicated Code — the FileEntity assembly (const entity = {...file.entity, ...baseFileTransformer(file), memberOf: …refMap.get…}) appears three times (file.ts GET + HEAD, files.ts). Extract into transformers/default.ts. Likewise the standardEntity block twice in crate.ts, and createForbiddenError('Access to this resource is restricted') five times.
- Data Clumps / missing type — resolveValidLicenses?: (opt: TransformerContext) => Promise<string[]> is re-declared inline in app.ts, entities.ts, files.ts, search.ts. Peer types (AccessTransformer, FileTransformer) live named in src/types/transformers.ts.
- Mysterious Name — FileEntity = StandardFile & StandardEntity where id is the file's but name/licences are the parent's; entity used as the variable name for it in file.ts; opt for a TransformerContext.
- Speculative Generality — (await resolveValidLicenses(...)) || [] (three sites); the declared return type is Promise<string[]>. The test only compiles via // @ts-expect-error (entities.test.ts:329), which hides the mismatch rather than fixing the signature.

###  Convention drift worth a look

- search.ts:66-70 places the await resolveValidLicenses call outside the try, so a throwing resolver bypasses the route's createInternalError() 500 — entities.ts/files.ts put it inside. It also mutates request.body.filters and silently overwrites a caller-supplied metadataLicenseId.
- files.ts:25 drops the destructuring default (fileTransformers = []) for inline fileTransformers || [], against the pattern in entities.ts/search.ts.
- Brace-less if returns in entity.ts:52-53 and crate.ts:94-95 versus braced everywhere else (Biome recommended doesn't enforce this).
- TESTING.md isolation — entities.test.ts uses a module-level mutable let hasLicense flipped inside tests, making the suite order-sensitive.

## Spec

### (a) Missing / partial

 /file doesn't implement the stated condition. Spec: "For endpoints /crate, /entity, /file: responds with 403 when access.metadata returned by accessTransformer function is false." src/routes/file.ts (both HEAD and GET) gates on !authorisedFile.access.content, not access.metadata — and FileAccessInfo has no metadata field at all (src/transformers/default.ts: "File metadata is
  always accessible - only content access is controlled"). Either the spec is wrong about /file, or the type needs a metadata flag. As written, the requirement is unmet and the endpoint gained a different restriction (content-gating) instead. Needs an author decision.

 No docs for the new public option. resolveValidLicenses is a new consumer-facing option in src/app.ts but appears nowhere in README.md, unlike every other option (accessTransformer, queryBuilderClass, etc.).

###  (b) Scope creep

  Breaking public API change to FileAccessTransformer. Not asked for anywhere in the spec. src/types/transformers.ts changes the parameter from StandardFile to FileEntity (StandardFile & StandardEntity), and AuthorisedFile gains all entity fields. Every consumer's fileAccessTransformer and the documented example in README.md:414 now receive a different shape. This is a
  major-semver change in a published library, driven only by the /file content check.

  /files response shape changed. src/routes/files.ts now spreads the raw Prisma entity (...dbFile.entity) into each returned file, so /files responses leak meta, createdAt, updatedAt, name, description, entityType, and both licence IDs. Spec asked only to "filter the search query".

### (c) Implemented but looks wrong

  /crate builds a malformed entity for the access transformer. src/routes/crate.ts (both routes) uses memberOf: { id: entity.memberOf || '', name: '' } — never null and never a resolved name — instead of baseEntityTransformer + resolveEntityReferences as /entity does. A consumer transformer keying on memberOf/rootCollection (the common case) will see {id:'',name:''} for
  top-level entities and make wrong access decisions.

  Falsy vs false. Spec says "when access.metadata ... is false"; all three sites use !access.metadata, so undefined also 403s. Probably fine, but it is stricter than specified.

  /search licence filter likely matches nothing. filters.metadataLicenseId is pushed through queryBuilder.buildQuery as terms: { metadataLicenseId: [...] } on a raw field; every other exact-match field in the codebase uses .keyword (queryBuilder.ts:31) and metadataLicenseId is absent from the index mapping in src/test/integration.setup.ts. Under dynamic mapping it becomes
  analysed text, so URL licence values won't match — filtering the user out of all results. Only unit tests (mocked OpenSearch) cover this.

  /search clobbers caller filters and mutates request.body.filters in place; a caller-supplied metadataLicenseId filter is silently replaced rather than intersected.

File: src/routes/entities.test.ts (hunk 6 above)
Comment (alvinsw) [reply src/routes/entities.test.ts]: The @ts-expect-error is required to test if the implementation of resolveValidLicenses function returns undefined or null, see below.

File: src/routes/entities.ts (hunk 7 above)
Comment (alvinsw) [reply src/routes/entities.ts]: It is just an defensive approach as the library can be used as JS library and it is better to provide a sane dafault.

File: src/routes/crate.ts (hunk 8 above)
Comment (alvinsw) [reply src/routes/crate.ts]: To keep it simple, as they are not really used yet, and you need to recompute the result in the resolveEntityReferences function. We only need to check the access there by calling the accessTransformer.

File: src/routes/crate.ts (hunk 9 above)
Comment (alvinsw) [reply src/routes/crate.ts]: as above

File: src/transformers/default.ts (hunk 13 above)
Comment (alvinsw) [reply src/transformers/default.ts]: The file access transformer needs to check the the `metadataLicenseId` too, which is not defined in the StandardFile.

File: src/routes/files.ts (hunk 10 above)
Comment (alvinsw) [reply src/routes/files.ts]: Because originally it only passes standardFile which does not have the metadataLicenseId and contentLicenseId required to determine access in the fileAccessTransformer function.

File: src/routes/search.ts (hunk 11 above)
Comment (alvinsw) [reply src/routes/search.ts]: What do you mean by "the rationale for this"? We can just discuss them here.
If the metadata access for files is restricted then such files should should be shown in the search result.

File: src/transformers/default.ts (hunk 12 above)
Comment (alvinsw) [reply src/transformers/default.ts]: As mentioned elsewhere, StandardFile type doesn't have all the required fields.

File: src/routes/entities.test.ts (hunk 6 above)
Comment (johnf) [reply src/routes/entities.test.ts]: Defining the function as

```ts
  async function resolveValidLicenses() {
    if (hasLicense) {
      return ['https://creativecommons.org/licenses/by/4.0/'];
    }

    return [];
  }
```
Should remove the need for the error comment

File: src/routes/entities.ts (hunk 7 above)
Comment (johnf) [reply src/routes/entities.ts]: Hmm, I'm not convinced. This means that if we are provided with a function that has a bug i.e. it has a logic error and doesn't return anything, then we return unexpected data instead.

If you want to be defensive, then throw an error if the function does the wrong thing.

It is up to the caller to honour the contract. If they don't, that's a bug in their code that we are hiding from them.

File: src/routes/crate.ts (hunk 8 above)
Comment (johnf) [reply src/routes/crate.ts]: But then we are passing invalid data into `accessTransformer`. What if they wanted to restrict access based on the `name` of `memberOf`?

File: src/routes/search.ts (hunk 11 above)
Comment (johnf) [reply src/routes/search.ts]: I think it is really easy for someone to assume that `File/MediaObject` will always be gated by the `contentMetadataLicense`. That is not the approach we are taking because we clearly distinguish between metadata for a file and the file itself.

This needs to be clearly documented.

File: src/routes/entities.test.ts (hunk 6 above)
Comment (alvinsw) [reply src/routes/entities.test.ts]: That would fail the coverage test (branch section)
